xc/xenops_server_xen.ml

```diff
@@ -2574,6 +2580,25 @@ module Actions = struct
 					| _ -> ()
 			) (DB.read vm)
 
+	let maybe_update_pv_drivers_detected ~xc ~xs domid path =
+		let vm = get_uuid ~xc domid |> Uuidm.to_string in
+		Opt.iter
+			(function { VmExtra.persistent; non_persistent } ->
+				if not non_persistent.VmExtra.pv_drivers_detected then begin
+					(* If the new value for this device is 4 then PV drivers are present *)
+					try
+						let value = xs.Xs.read path in
+						if value = "4" (* connected *) then begin
+							let non_persistent = { non_persistent with VmExtra.pv_drivers_detected = true } in
+							debug "VM = %s; found PV driver evidence on %s (value = %s)" vm path value;
+							DB.write vm { VmExtra.persistent; non_persistent }
+						end
+					with Xs_protocol.Enoent _ ->
+						warn "Watch event on %s fired but couldn't read from it" path;
```

Comment:
That's fine. We're just auditing new additions to the logging. You have thought it through which is enough :)